sklearn/linear_model/huber.py

```diff
@@ -0,0 +1,347 @@
+        grad[:n_features] = (
+            2 * safe_sparse_dot(non_outliers, X_non_outliers) / sigma)
+    else:
+        grad[:n_features] = (
+            2 * safe_sparse_dot(weighted_non_outliers, X_non_outliers) / sigma)
+
+    # Gradient due to the linear loss.
+    outliers_pos = np.logical_and(linear_loss >= 0, outliers_mask)
+    outliers_neg = np.logical_and(linear_loss < 0, outliers_mask)
+    num_outliers_pos = np.count_nonzero(outliers_pos)
+    num_outliers_neg = num_outliers - num_outliers_pos
+    X_outliers_pos = _axis0_safe_slice(X, outliers_pos, num_outliers_pos)
+    X_outliers_neg = _axis0_safe_slice(X, outliers_neg, num_outliers_neg)
+
+    if sample_weight is None:
+
+        # Summing along any axis of a matrix returns a matrix object.
+        outliers_pos_sum = np.squeeze(np.array(X_outliers_pos.sum(axis=0)))
+        outliers_neg_sum = np.squeeze(np.array(X_outliers_neg.sum(axis=0)))
+
+        grad[:n_features] -= 2 * epsilon * outliers_pos_sum
+        grad[:n_features] += 2 * epsilon * outliers_neg_sum
+
+    else:
+        sw_outliers_pos = sample_weight[outliers_pos]
+        sw_outliers_neg = sample_weight[outliers_neg]
+
+        if X_sparse:
+            weighted_sum = (
+                _dia_matrix(sw_outliers_pos) *
+                X_outliers_pos).sum(axis=0)
+            weighted_sum = np.squeeze(np.array(weighted_sum))
+            grad[:n_features] -= 2 * epsilon * weighted_sum
+
+            weighted_sum = (
+                _dia_matrix(sw_outliers_neg) *
+                X_outliers_neg).sum(axis=0)
+            weighted_sum = np.squeeze(np.array(weighted_sum))
+            grad[:n_features] += 2 * epsilon * weighted_sum
+        else:
+            grad[:n_features] -= 2 * epsilon * (
+                np.dot(sw_outliers_pos, X_outliers_pos))
+            grad[:n_features] += 2 * epsilon * (
+                np.dot(sw_outliers_neg, X_outliers_neg))
+
+    # Gradient due to the penalty.
+    grad[:n_features] += alpha * 2 * w
+
+    # Gradient due to sigma.
+    grad[-1] = n_samples
+    grad[-1] -= n_outliers * epsilon**2
+    grad[-1] -= squared_loss / sigma
+
+    # Gradient due to the intercept.
+    if fit_intercept:
+        if sample_weight is None:
+            grad[-2] = -2 * np.sum(non_outliers) / sigma
+            grad[-2] -= 2 * epsilon * np.sum(outliers_pos)
+            grad[-2] += 2 * epsilon * np.sum(outliers_neg)
+        else:
+            grad[-2] = -2 * np.sum(weighted_non_outliers) / sigma
+            grad[-2] -= 2 * epsilon * np.sum(sw_outliers_pos)
+            grad[-2] += 2 * epsilon * np.sum(sw_outliers_neg)
+
+    return n_samples * sigma + squared_loss + outlier_loss + alpha * np.dot(w, w), grad
+
+
+class HuberRegressor(LinearModel, RegressorMixin, BaseEstimator):
+    """
+    Linear regression model that is robust to outliers.
+
+    The Huber Regressor optimizes the squared loss for the samples where
+    |y - X'w / exp(sigma)| > epsilon and the linear loss for the samples
```

Comment:
Yes, I think we should call it absolute loss / error. consistent with metrics.mean_absolute_error as well.